### ISSUE TITLE
FONT_ANNOT_PRIMARY of colorbar not respected in subplots

### ISSUE THREAD

**Issue by hzfmer** [GenericMappingTools/gmt#4978]
**Description of the problem**

When plotting colorbars in `subplot`s, the font size of colorbar labels cannot be changed by `FONT_ANNOT_PRIMARY` or even `FONT`.

```
range=-R0/200/0/100
gmt begin test png
gmt subplot begin 2x1 -Fs6c/4.5c -M0/0c -A+o-0.4c/-0.0c
  gmt subplot set 0 -A'(a)'
  gmt colorbar ${range} -Chot -DjMR+o-0.4c/0+m+ef -I0.3  --FONT_ANNOT_PRIMARY=5p

  gmt subplot set 1 -A'(b)'
  gmt basemap -JX? ${range} -Btblr
  gmt colorbar -Chot -DjMR+o-0.4c/0+m+ef -I0.3  --FONT_ANNOT_PRIMARY=1p

gmt subplot end
gmt end
```

**Full error message**

It doesn't trigger an error, just not working as expected.

**Actual outcome**
![image](https://user-images.githubusercontent.com/28425640/111424510-1edb0100-86af-11eb-9929-98b0f6c69eb7.png)

**Expected outcome**
Setting to something like `1p` should generate small texts. Also, whether `basemap` exists or not should not change the behavior (In this case, when no `basemap` or anything else creating a frame, the parameter `FONT_ANNOT_PRIMARY` does show effect)

However, just reversing the order of these two subplots makes it different:
```
range=-R0/200/0/100
gmt begin test png
gmt subplot begin 2x1 -Fs3c/2c -M0/0c -A+o-0.4c/-0.0c
  gmt subplot set 0 -A'(a)'
  gmt basemap -JX? ${range} -Btblr
  gmt colorbar -Chot -DjMR+o-0.4c/0+m+ef -I0.3  --FONT_ANNOT_PRIMARY=1p

  gmt subplot set 1 -A'(b)'
  gmt colorbar ${range} -Chot -DjMR+o-0.4c/0+m+ef -I0.3  --FONT_ANNOT_PRIMARY=5p

gmt subplot end
gmt end
```
![image](https://user-images.githubusercontent.com/28425640/111425024-c5270680-86af-11eb-89ab-084c494a3ac3.png)

Furthermore, this function works fine in GMT 6.1.1.

**System information**

* Operating system:
MacOS Mojave
* GMT version (`gmt --version`)
6.2.0_16e81ff_2021.03.16

**Comment by PaulWessel** [comment]
@meghanrjones, looks like the auto-scaling happens here even if --FONT_ANNOT_PRIMARY=5p is set:

`colorbar [DEBUG]: Computed primary annotation font size: 7.76267 p  Dimension scaling: 0.776267`

suggesting the nominal font size is 10p prior to the size scaling.

**Comment by meghanrjones** [comment]
> @meghanrjones, looks like the auto-scaling happens here even if --FONT_ANNOT_PRIMARY=5p is set:
> 
> `colorbar [DEBUG]: Computed primary annotation font size: 7.76267 p Dimension scaling: 0.776267`
> 
> suggesting the nominal font size is 10p prior to the size scaling.

Yes, I will debug to figure out why that is happening.

**Comment by meghanrjones** [comment]
> @meghanrjones, looks like the auto-scaling happens here even if --FONT_ANNOT_PRIMARY=5p is set:
> 
> `colorbar [DEBUG]: Computed primary annotation font size: 7.76267 p Dimension scaling: 0.776267`
> 
> suggesting the nominal font size is 10p prior to the size scaling.

That message is printed regardless of whether font_annot_primary is reset:

```
	GMT_Report (GMT->parent, GMT_MSG_DEBUG, "Computed primary annotation font size: %g p  Dimension scaling: %g\n", fontsize, scale);

	/* Only apply the automatic scaling to items NOT specifically set via a --PAR=value option */

	if (gmt_M_is_dnan (GMT->current.setting.font_annot[GMT_PRIMARY].size)) {
		GMT->current.setting.font_annot[GMT_PRIMARY].size = fontsize;
		if (conf_update) GMT_keyword_updated[GMTCASE_FONT_ANNOT_PRIMARY] = true;
	}
```

I'll fix that and am still working on the bug.

**Comment by meghanrjones** [comment]
@PaulWessel, the problem is that ``parse`` calls ``gmt_auto_offsets_for_colorbar`` which, if there's a session file with frame attributes, calls ``gmtinit_conf_modern_override``, which resets anything set by ``GMT_Parse_Common``

Here's the relevant lines from the start of psscale.c:
```
	if (GMT_Parse_Common (API, THIS_MODULE_OPTIONS, options)) Return (API->error);
	Ctrl = New_Ctrl (GMT);	/* Allocate and initialize a new control structure */
	if ((error = parse (GMT, Ctrl, options)) != 0) Return (error);
```

I think any GMT defaults set during the session should not be reset, for example this should have a 20p colorbar primary annotation:

```
gmt begin test_font
gmt set FONT_ANNOT_PRIMARY=20p
gmt basemap -JX10c -R0/200/0/100 -Bafg -Vd
gmt colorbar -Chot -DjMR+o-0.4c/0+m+ef -I0.3 -Vd
gmt end show
```

Also, just to be clear this is not a exclusively subplot problem, but will happen with any modern-mode auto-positioned colormap:

```
gmt begin test1
gmt basemap -JX10c -R0/200/0/100 -Ba -Vd
gmt colorbar -Chot -DjMR+o-0.4c/0+m+ef -I0.3 --FONT_ANNOT_PRIMARY=20p -Vd
gmt end show
```

**Comment by meghanrjones** [comment]
> I wonder if we pass _options_ into gmt_auto_offsets_for_colorbar (...) then perhaps after gmtinit_conf_modern_override (GMT) is called at the end of it we can re-process just the **--PAR**=_value_ arguments? That should replace the auto with actual values. I.e., at the very end of _gmt_auto_offsets_for_colorbar_, do something like this
> 
> ```
> void gmt_auto_offsets_for_colorbar (struct GMT_CTRL *GMT, double offset[], int justify, struct GMT_OPTION *options) {
> ...
> 
> for (opt = options; opt; opt = opt->next) {
>   if (opt->option != ' -') continue;   /* Not a parameter setting */
>   c = strchr (opt->arg, '=');
>   c[0] = '\0';  /* Remove = */
>   n_errors += gmtlib_setparameter (GMT, opt->arg, &c[1], false);
> }
> ```

That wouldn't fix this case (do you think the FONT_ANNOT_PRIMARY from earlier in the session shouldn't apply to colorbar?):

```
gmt begin test_font
gmt set FONT_ANNOT_PRIMARY=20p
gmt basemap -JX10c -R0/200/0/100 -Bafg -Vd
gmt colorbar -Chot -DjMR+o-0.4c/0+m+ef -I0.3 -Vd
gmt end show
```

**Comment by PaulWessel** [comment]
Right.  So there may be a gmt.conf that has non-auto settings for fonts, and there may be --PAR-value incoming.  I need coffee and breakfast first I think.

**Comment by meghanrjones** [comment]
I think using ``gmt_reload_settings (GMT)`` rather than ``gmtinit_conf_modern_override (GMT)`` might work. Checking it now.

**Comment by meghanrjones** [comment]
@hzfmer, your examples should work properly now after PR #4982.

**Comment by hzfmer** [comment]
> @hzfmer, your examples should work properly now after PR #4982.

Thank you @MarceloBanik and @PaulWessel  for your impressively quick fix. I used Anaconda to install GMT so I cannot verify how it works now, but I will try it once the conda package is updated.

**Comment by seisman** [comment]
It looks like a bug that affects many plots. 

@weiji14 Could you please update the conda-forge dev version when you have time?

**Comment by weiji14** [comment]
Sure, I can trigger a new version build. Give me a few minutes.

**Comment by weiji14** [comment]
@hzfmer, you can now install the latest GMT dev version using `conda install -c conda-forge/label/dev gmt`. That will pull in GMT version `6.2.0.dev12+922af84` which contains the fix at #4982.

**Comment by hzfmer** [comment]
> @hzfmer, you can now install the latest GMT dev version using `conda install -c conda-forge/label/dev gmt`. That will pull in GMT version `6.2.0.dev12+922af84` which contains the fix at #4982.

It does work! Thank you all again!
